providers/common/ciphers/aes_gcm_s390x.inc

```diff
@@ -0,0 +1,327 @@
+    OPENSSL_cleanse(out, ctx->plat.s390x.mreslen);
+    ctx->iv_state = IV_STATE_FINISHED;
+
+    if (ctx->enc) {
+        ctx->taglen = 16;
+        memcpy(tag, kma->t.b, ctx->taglen);
+    } else {
+        if (ctx->taglen < 0)
+            return 0;
+
+        if (CRYPTO_memcmp(tag, kma->t.b, ctx->taglen) != 0)
+            return 0;
+    }
+    return 1;
+}
+
+/*
+ * in_len = length of PlainText & CipherText
+ * For encryption:
+ *   in = Plaintext
+ *   out = CipherText + tag
+ * For decryption:
+ *   in = CipherText + Tag
+ *   out = returned plaintext
+ * tag is the calculated Authentication Tag (with a size of 16 bytes)
+ */
+static int s390x_aes_gcm_one_shot(PROV_AES_GCM_KEY *ctx,
+                                  unsigned char *aad, size_t aad_len,
+                                  const unsigned char *in, size_t in_len,
+                                  unsigned char *out,
+                                  unsigned char *tag, size_t taglen)
+{
+    S390X_KMA_PARAMS *kma = &ctx->plat.s390x.param.kma;
+
+    kma->taadl = aad_len << 3;
+    kma->tpcl = in_len << 3;
+    s390x_kma(aad, ctx->tls_aad_len, in, in_len, out,
+              ctx->plat.s390x.fc | S390X_KMA_LAAD | S390X_KMA_LPC, kma);
+    memcpy(tag, kma->t.b, taglen);
+    return 1;
+}
+
+/*
+ * Process additional authenticated data. Returns 1 on success. Code is
+ * big-endian.
+ */
+static int s390x_aes_gcm_aad_update(PROV_AES_GCM_KEY *ctx,
+                                    const unsigned char *aad, size_t len)
+{
+    S390X_KMA_PARAMS *kma = &ctx->plat.s390x.param.kma;
+    unsigned long long alen;
+    int n, rem;
+
+    /* If already processed pt/ct then error */
+    if (kma->tpcl != 0)
+        return 0;
+
+    /* update the total aad length */
+    alen = kma->taadl + len;
+    if (alen > (U64(1) << 61) || (sizeof(len) == 8 && alen < len))
+        return 0;
+    kma->taadl = alen;
+
+    /* check if there is any existing aad data from a previous add */
+    n = ctx->plat.s390x.areslen;
+    if (n) {
+        /* add additional data to a buffer until it has 16 bytes */
+        while (n && len) {
+            ctx->plat.s390x.ares[n] = *aad;
+            ++aad;
+            --len;
+            n = (n + 1) & 0xf;
+            /* ctx->ares contains a complete block if offset has wrapped around */
+            if (n == 0) {
+                s390x_kma(ctx->plat.s390x.ares, 16, NULL, 0, NULL,
+                          ctx->plat.s390x.fc, kma);
+                ctx->plat.s390x.fc |= S390X_KMA_HS;
+                break;
+            }
+        }
+        ctx->plat.s390x.areslen = n;
+    }
+
+    /* Add any remaining 16 byte blocks (128 bit each) */
+    len &= ~(size_t)0xf;
+    if (len) {
+        s390x_kma(aad, len, NULL, 0, NULL, ctx->plat.s390x.fc, kma);
+        ctx->plat.s390x.fc |= S390X_KMA_HS;
+        aad += len;
+    }
+
+    /* If there are leftover bytes (< 128 bits) save them for next time */
+    rem = len & 0xf;
```

Comment:
This statement has to stay where it was in my original submission.

Reply:
You are right - Not sure why I moved it.